Feature/demean accelerated

schroedk opened the pull request:
This PR implements a simple version of the Iron-Tucks acceleration.

s3alfisc commented:
@schroedk just rebased / merged changes from master in here (new features plus I moved to using a pixi toml, plus the dev env no longer installs on windows due to compatibility challenges). 

As before, you can build the Rust bindings by typing
```
pixi r -e dev maturin-develop
```
and run the benchmarks via 
```
pixi run -e dev pytest  tests/test_demean.py::test_demean_complex_fixed_effects
```

s3alfisc commented:
Benchmarks of the accelerated vs regular rust vs fixest and FixedEffectsModels.jl: 

<img width="5360" height="3564" alt="image" src="https://github.com/user-attachments/assets/be81ce6f-9778-4ce0-b453-aa44b009c482" />

s3alfisc commented:
Yes, it's here: https://github.com/s3alfisc/fixest_benchmarks

Hope I documented the setup well, but I think clone + just + task runner should get you started. 

It's the OLS benchmarks for the hard problem that are relevant. 

Note: @grantmcdermott mentioned the other day there might be a minor issue with the benchmarks (though I don't know what exactly) so best to take it with a grain of salt (though I couldn't spot it, everything looked ok to me).

s3alfisc commented:
Wait it looks like I didn't push my local changes including the just setup. One sec

s3alfisc commented:
It's in the justfile branch on the remote 😅

https://github.com/s3alfisc/fixest_benchmarks

Requirements: global R and Julia installations. Just. 

Then type 

just setup 

To install all package deps as well as python (in local env). 

Then just bench-ols for benchmarks.

s3alfisc commented:
Is this confirmed? 👀 😄 🚀 

<img width="591" height="69" alt="image" src="https://github.com/user-attachments/assets/b4cddfdf-d1da-4778-adfb-9a65bc1d0a61" />

s3alfisc commented on pyfixest/core/_core_impl.pyi:
```diff
@@ -20,3 +20,11 @@ def _count_fixef_fully_nested_all_rs(
+    x: NDArray[np.float64],
+    flist: NDArray[np.uint64],
+    weights: NDArray[np.float64],
+    tol: float = 1e-08,
```
```suggestion
    tol: float = 1e-06,
```

s3alfisc commented on pyfixest/core/demean_accelerated.py:
```diff
@@ -0,0 +1,79 @@
+    x: NDArray[np.float64],
+    flist: NDArray[np.uint64],
+    weights: NDArray[np.float64],
+    tol: float = 1e-08,
```
```suggestion
    tol: float = 1e-06,
```

schroedk commented:
> Awesome, I will start my review then! I will try to be fast, but also have a [big PR](https://github.com/py-econometrics/pyfixest/pull/1118) from @leostimpfle to review thoroughly + all the Rust code will be a bit of a challenge for me 😄
> 
> On your questions, I'd say:
> 
> 1. Yes, let's delete the benchmark code? I will push some changes to the [fixest_benchmark](https://github.com/s3alfisc/fixest_benchmarks) repo that I have been working on going forward, we can use it to benchmark the routines?
> 2. This is super awesome! For the numba implementation - I guess I would keep it for now? I haven't fully decided what to do with the numba code once we make the switch and make the rust backend the default - option 1 is to allow users to choose a numba backend, option 2 would be to delete all numba code. Do you happen have an opinion on this?
> 3. I would probably delete it? Or is there any use in keeping it for testing / as a reference?
> 4. Yes, squash commit makes sense to me 👍

1. deleted
2. I will keep it for now, feols already uses the rust version. There is no other usage (beside tests) of the numba code, so one can rid of it any time and re-export the rust backed version for backwards compatibility. I cannot think of a reason to keep the numba backend in the middle/long run, so I would go with Option 2.
3. I deleted it and rewired the python bindings to the accelerated version (commit 
6c091d7) -> no breaking change in the python API
4. I created a smaller more meaningful set of commits, so you could also step through the commits for the review. If you like, we can squash to one commit for merge.

s3alfisc commented:
If I understand it correctly, the algorithm is set up to operate in "coef space" - this means that we have estimated fixed effects at the end of the demeaning routine. Would it be possible to return these fixed effects? In these fixed effects models, we are usually not interested in them, but we might need them to conduct predictions, which in turn is needed to compute marginal effects quickly. This is more or less how the R `margins` package does it, and Stata and R and Pythons `marginal effects` too. From the [margins paper](https://cran.r-project.org/web/packages/margins/vignettes/TechnicalDetails.pdf): 
<img width="511" height="96" alt="image" src="https://github.com/user-attachments/assets/d8eef382-48f9-4b9a-95ee-aaa83909a333" />
<img width="511" height="199" alt="image" src="https://github.com/user-attachments/assets/c006f6ed-397b-49b9-aeda-290b976bf0c5" />

In pyfixest, we currently get the estimated fixed effects post hoc by solving a sparse system via LSMR (see [here](https://github.com/py-econometrics/pyfixest/blob/8d07f0b36b969194ab52ef737833ff1d039f71c3/pyfixest/estimation/feols_.py#L2005)), which I think we could skip now?

s3alfisc commented:
> The bigger problem is, how to add it to the python part?

Fixest has a Boolean argument "reorder_coef" - I think we could just add it to the python interface and then pass it to the Rust code? For the non-accelerated algos, it would just not do anything. Maybe I am missing something?
